### PR TITLE
E208: Corrected file mode checking

### PR DESCRIPTION
**Pull request by ssbarnea**
Takes care of incorrect Ansible documentation which misleads users about support for `mode: preserve` on modules which do not accept it.

Fixes: #1007

### CODE REVIEW

**Comment by sivel** [comment]
We've talked about this rule a little bit and aren't sure this is beneficial.  The original change that necessitated this rule, has been reverted, and we have no intention on adding a change back that affects the result like the previous change did.

I question what this rule is attempting to achieve at this point.

If you think it should be kept, I'm open to hearing the reasoning to understand your POV.  However, I don't think we should really worry with telling people about `preserve`.

Generally speaking, I feel like this rule isn't necessarily "best practice". The problem with ansible-lint, compared to something like pycodestyle for example, is that it's hard to understand intent and need.  While it may be a good idea to use `mode` in some situations, it's not something that is required everywhere.  So making it a rule that people have to disregard in many situations seems like busy work for no benefit.

We've discussed that while we could have many pedantic rules, many should probably be disabled by default.  And maybe there could be a CLI flag called `--strict` or `--pedantic` that enables them all.

**Comment by nitzmahone** [comment]
+1 to @sivel 's comments. A pattern I've liked in other linters is user-definable rule set files rather than (or in addition to) skiplists. The tool defaults to one of the shipped rule sets, but that default is configurable, and users can trivially supply their own rulesets as well (or copy one of the builtins and augment). Something like that would accommodate the "off by default" behavior that we'd like to see for the previous new rule- it's not part of the default rule set, but could be in a `pedantic` or dynamic `all` ruleset (and it's there for users to include in their own rulesets should they choose).

**Comment by ssbarnea** [comment]
Regardless the fact that the CVE related mode change was reverted the subject of missing `mode:` is as valid as possible, as Ansible behavior is still not documented when the mode is not mentioned.

The documentation about "mode" which is also false, advertising unsupported values like "preserve". Only 3/11 modules supporting preserve in reality.

I seen a patch being merged to remove the mention of preserve but this was not backported to all previous versions of ansible, which people may still use. Some of them may even have invalid values in their code because the code was not effectively tested.

IMHO, I would be a serious mistake to ignore the `mode` range of issues.

Even if documentation would be updated to mention clearly what is the file mode behavior for all 11 modules when mode is not mentioned, we still need this rule to underline the wrong usage of "preserve" on those 8 modules that do not allow it. Such a documentation mistake that was online for more than an year does not go away over night. I still see preserve being advertised on both 2.8 and 2.9.

**Comment by sivel** [comment]
Sure, then I suppose at that level, the rule becomes only a way to make sure people are _not_ using `preserve` on a subset of modules, not giving them instruction that they _need_ to use `mode`.

**Comment by ssbarnea** [comment]
> Sure, then I suppose at that level, the rule becomes only a way to make sure people are _not_ using `preserve` on a subset of modules, not giving them instruction that they _need_ to use `mode`.

I am more than happy to repurpose the rule to make it only spot incorrect use of preserve but can we first get the mode behavior documented for cases when it is not mentioned?

At this moment the documentation does not document this:

![https://sbarnea.com/ss/Screen-Shot-2020-09-03-18-59-38.44.png][]

Having examples on how to use it is good, but not mentioning what happens when you don't is clearly an issue, a potential security issue because created files may contain credentials.

The expectation of the most people I asked was the umode would be used, but apparently that is not the case.

**Comment by sivel** [comment]
This was just merged https://github.com/ansible/ansible/pull/71486

**Comment by ssbarnea** [comment]
I am aware about the partial removal of `preseve` from documentation, but that is only half of the story. We still do not have documentation about which file modes ansible will use for creating these files/directories.

I asked @acozine for help here, with bit of luck we may sort this issue.

**Comment by ssbarnea** [comment]
Regarding the proposal of avoiding new rules or specific groups of rules, I raised https://github.com/ansible/ansible-lint/issues/1031 proposal. Please review it.

File: lib/ansiblelint/rules/MissingFilePermissionsRule.py [hunk 1]
```diff
@@ -19,45 +19,58 @@
 # THE SOFTWARE.
 from ansiblelint.rules import AnsibleLintRule
 
+# Despite documentation mentioning 'preserve' only these modules support it:
+_modules_with_preserve = (
+    'copy',
+    'ini_file',
```
Comment (Sacrome) [inline lib/ansiblelint/rules/MissingFilePermissionsRule.py:25]: Looking at the code of `ini_file` I don't see any references to the argument `mode: preserve`. 
I suspect that only the modules `copy` and `template` have this functionality.

Comment (Sacrome) [reply lib/ansiblelint/rules/MissingFilePermissionsRule.py:25]: Comment by @sivel confirms my claim: https://github.com/ansible/ansible/issues/56839#issuecomment-680082091
> @geerlingguy that is correct. Per my comment above, `preserve` only applies to `copy` and by extension `template`. Someone incorrectly moved some docs into the shared fragment which was wrong, no one has submitted a fix for the docs issue as of yet.

**Review by Sacrome — CHANGES_REQUESTED** [review]

File: lib/ansiblelint/rules/MissingFilePermissionsRule.py (hunk 1 above)
Comment (ssbarnea) [reply lib/ansiblelint/rules/MissingFilePermissionsRule.py]: Thanks for spotting the bug. I fixed it and also extended the tests. Please take another look, hopefully we will do it right this time.

**Review by Sacrome — APPROVED** [review]

**Comment by Sacrome** [comment]
@ssbarnea Thanks for updating the linter! 
I've already approved your PR, but left a small comment, please take a look at it and see if you agree 😄

**Comment by odyssey4me** [comment]
re: ``preserve``
It seems to me that if a module doesn't accept/use a parameter then that's a bug which should be solved in ansible itself - not in a linting tool.

**Comment by Sacrome** [comment]
> re: `preserve`
> It seems to me that if a module doesn't accept/use a parameter then that's a bug which should be solved in ansible itself - not in a linting tool.

The documentation has actually been fixed in Ansible v2.10 and forward, with two PRs open for backporting this to v2.9 and v2.8. The linter however, incorrectly suggests `mode: preserve` on a number of modules. Which is why this PR exists.

**Comment by Sacrome** [comment]
Note that as of Ansible v2.9.13 the behaviour of the `mode` parameter has been changed, reverting the fix for CVE-2020-1736. 

https://github.com/ansible/ansible/blob/stable-2.9/changelogs/CHANGELOG-v2.9.rst#security-fixes

**Comment by ssbarnea** [comment]
> Note that as of Ansible v2.9.13 the behaviour of the `mode` parameter has been changed, reverting the fix for CVE-2020-1736.
> 
> [https://github.com/ansible/ansible/blob/stable-2.9/changelogs/CHANGELOG-v2.9.rst#security-fixes](https://github.com/ansible/ansible/blob/stable-2.9/changelogs/CHANGELOG-v2.9.rst?rgh-link-date=2020-09-07T12%3A07%3A20Z#security-fixes)

E208 has almost nothing to do with the mentioned CVE. It has to do with the fact that with any version of Ansible, if you do not specify a value for "mode", you will end-up having unpredictable file permissions. The final result does not only depend on the version of Ansible but also on the operating system and the connection method. That is issue that this rule is trying to warn about and this is why is tagged with `unpredictability`. Arguably we could also use `security` as not being explicit about it, is also a potential security issue.

Please note that rule was tagged as experimental, this means it will no longer produce errors by default (they will only be warnings).

**Comment by ssbarnea** [comment]
> re: `preserve`
> It seems to me that if a module doesn't accept/use a parameter then that's a bug which should be solved in ansible itself - not in a linting tool.

Ansible itself is moving at a very slow pace and they struggle even to fix documentation errors. For example, even if we all discovered that preserve was documented on all 11 modules when only 3 were supporting it week ago, preserve was removed only from devel. Guess what, >95% of ansible users are reading documentation of 2.7, 2.8 and 2.9. For good reasons they read the same version of documentation as the version of the ansible they use in production.

As of today there are not even opened PR to backport the documentation fixes. These being said, that is very different from the linter, which can warn about both unsafe code. The linter is also a primary input tool that enables people to upgrade their roles/playbooks to newer versions of Ansible.

**Comment by Sacrome** [comment]
> > re: `preserve`
> > It seems to me that if a module doesn't accept/use a parameter then that's a bug which should be solved in ansible itself - not in a linting tool.
> 
> Ansible itself is moving at a very slow pace and they struggle even to fix documentation errors. For example, even if we all discovered that preserve was documented on all 11 modules when only 3 were supporting it week ago, preserve was removed only from devel. Guess what, >95% of ansible users are reading documentation of 2.7, 2.8 and 2.9. For good reasons they read the same version of documentation as the version of the ansible they use in production.
> 
> As of today there are not even opened PR to backport the documentation fixes. These being said, that is very different from the linter, which can warn about both unsafe code. The linter is also a primary input tool that enables people to upgrade their roles/playbooks to newer versions of Ansible.

Just to add to that;
I have actually opened up PR's last weekend for backporting the documentation to 2.8 and 2.9: 
https://github.com/ansible/ansible/pull/71632
https://github.com/ansible/ansible/pull/71633

But indeed, I'm still waiting for them to be reviewed and merged.

**Comment by Sacrome** [comment]
> > re: `preserve`
> > It seems to me that if a module doesn't accept/use a parameter then that's a bug which should be solved in ansible itself - not in a linting tool.
> 
> Ansible itself is moving at a very slow pace and they struggle even to fix documentation errors. For example, even if we all discovered that preserve was documented on all 11 modules when only 3 were supporting it week ago, preserve was removed only from devel. Guess what, >95% of ansible users are reading documentation of 2.7, 2.8 and 2.9. For good reasons they read the same version of documentation as the version of the ansible they use in production.
> 
> As of today there are not even opened PR to backport the documentation fixes. These being said, that is very different from the linter, which can warn about both unsafe code. The linter is also a primary input tool that enables people to upgrade their roles/playbooks to newer versions of Ansible.

Alright, thanks for looking into that. I found the changelog to be very sparse and wasn't aware what it was exactly - as no ticket or PR was mentioned. So I felt like mentioning it here might've been useful 😉

**Comment by Sacrome** [comment]
Alright, thanks for looking into that. I found the changelog to be very sparse and wasn't aware what it was exactly - as no ticket or PR was mentioned. So I felt like mentioning it here might've been useful 😉

> > Note that as of Ansible v2.9.13 the behaviour of the `mode` parameter has been changed, reverting the fix for CVE-2020-1736.
> > [https://github.com/ansible/ansible/blob/stable-2.9/changelogs/CHANGELOG-v2.9.rst#security-fixes](https://github.com/ansible/ansible/blob/stable-2.9/changelogs/CHANGELOG-v2.9.rst?rgh-link-date=2020-09-07T12%3A07%3A20Z#security-fixes)
> 
> E208 has almost nothing to do with the mentioned CVE. It has to do with the fact that with any version of Ansible, if you do not specify a value for "mode", you will end-up having unpredictable file permissions. The final result does not only depend on the version of Ansible but also on the operating system and the connection method. That is issue that this rule is trying to warn about and this is why is tagged with `unpredictability`. Arguably we could also use `security` as not being explicit about it, is also a potential security issue.
> 
> Please note that rule was tagged as experimental, this means it will no longer produce errors by default (they will only be warnings).

Alright, thanks for looking into that. I found the changelog to be very sparse and wasn't aware what it was exactly - as no ticket or PR was mentioned. So I felt like mentioning it here might've been useful 😉

File: lib/ansiblelint/rules/MissingFilePermissionsRule.py [hunk 2]
```diff
@@ -19,57 +19,77 @@
...
+_modules_with_preserve = (
+    'copy',
+    'template',
+)
+
 
 class MissingFilePermissionsRule(AnsibleLintRule):
     id = "208"
-    shortdesc = 'File permissions not mentioned'
+    shortdesc = 'File permissions not mentioned or incorrect'
```
Comment (webknjaz) [inline lib/ansiblelint/rules/MissingFilePermissionsRule.py:31]:
```suggestion
    shortdesc = 'File permissions unset or incorrect'
```

File: lib/ansiblelint/rules/MissingFilePermissionsRule.py [hunk 3]
```diff
@@ -19,57 +19,77 @@
 # THE SOFTWARE.
 from ansiblelint.rules import AnsibleLintRule
 
+# Despite documentation mentioning 'preserve' only these modules support it:
+_modules_with_preserve = (
```
Comment (webknjaz) [inline lib/ansiblelint/rules/MissingFilePermissionsRule.py:23]: It looks like it's semantically better to use set and not a tuple here.

File: lib/ansiblelint/rules/MissingFilePermissionsRule.py [hunk 4]
```diff
@@ -19,57 +19,77 @@
...
     id = "208"
-    shortdesc = 'File permissions not mentioned'
+    shortdesc = 'File permissions not mentioned or incorrect'
     description = (
-        "Missing mode parameter can cause unexpected file permissions based "
-        "on version of Ansible being used. Be explicit, or if you still "
-        "want the default behavior you can use ``mode: preserve`` to avoid "
-        "hitting this rule. See "
-        "https://github.com/ansible/ansible/issues/71200"
+        f"Missing or unsupported mode parameter can cause unexpected file "
```
Comment (webknjaz) [inline lib/ansiblelint/rules/MissingFilePermissionsRule.py:33]: f-prefix is only necessary on lines with interpolation
```suggestion
        "Missing or unsupported mode parameter can cause unexpected file "
```

File: lib/ansiblelint/rules/MissingFilePermissionsRule.py [hunk 5]
```diff
@@ -19,57 +19,77 @@
...
-        "Missing mode parameter can cause unexpected file permissions based "
-        "on version of Ansible being used. Be explicit, or if you still "
-        "want the default behavior you can use ``mode: preserve`` to avoid "
-        "hitting this rule. See "
-        "https://github.com/ansible/ansible/issues/71200"
+        "Missing or unsupported mode parameter can cause unexpected file "
+        f"permissions based "
+        f"on version of Ansible being used. Be explicit, like ``mode: 0644`` to "
+        f"avoid hitting this rule. Special ``preserve`` value is accepted "
+        f"only by {', '.join(_modules_with_preserve)} modules. "
```
Comment (webknjaz) [inline lib/ansiblelint/rules/MissingFilePermissionsRule.py:37]: Only thing line needs an f prefix

File: lib/ansiblelint/rules/MissingFilePermissionsRule.py [hunk 6]
```diff
@@ -19,57 +19,77 @@
...
+        if module in self._modules_with_create:
+            create = task["action"].get("create", self._modules_with_create[module])
+            return create and mode is None
+
         # A file that doesn't exist cannot have a mode
         if task['action'].get('state', None) == "absent":
             return False
 
-        # A symlink always has mode 0o777
+        # A symlink always has mode 0777
```
Comment (webknjaz) [inline lib/ansiblelint/rules/MissingFilePermissionsRule.py:80]: This change is misleading. Before it, it was clearly octal, now it's not

File: lib/ansiblelint/rules/MissingFilePermissionsRule.py [hunk 7]
```diff
@@ -19,57 +19,77 @@
...
-    _modules_with_create = (
-        'blockinfile',
-        'ini_file',
-        'lineinfile'
-    )
+    _modules_with_create = {
+        'blockinfile': False,
+        'ini_file': True,
+        'lineinfile': False,
+        'htpasswd': True,
```
Comment (rafaelfolco) [inline lib/ansiblelint/rules/MissingFilePermissionsRule.py:58]: ok htpasswd apprently supports preserve from 2.6: https://docs.ansible.com/ansible/latest/modules/htpasswd_module.html#parameter-mode

don't you keep these lists in alphabetical order ?

**Review by rafaelfolco — APPROVED** [review]
lgtm

File: lib/ansiblelint/rules/MissingFilePermissionsRule.py (hunk 7 above)
Comment (Sacrome) [reply lib/ansiblelint/rules/MissingFilePermissionsRule.py:58]: The documentation of Ansible 2.9 is incorrect. The `htpasswd` module does not support `preserve`. 
I have verified this myself and created PRs for fixing the docs. You can read more on this problem here: https://github.com/ansible/ansible/issues/56839

**Review by rafaelfolco — APPROVED** [review]
lgtm

**Comment by gundalow** [comment]
What's needed to progress this?

**Review by sivel — APPROVED** [review]

**Review by gundalow — APPROVED** [review]

**Comment by bbaassssiiee** [comment]
I wonder what the correct permission would be for `unarchive:`? Is it the top-level directory? I always thought if `mode:` was undefined that it was governed by the umask of the use connecting.

**Comment by ssbarnea** [comment]
Unarchive is quite tricky because based on archive type it may or may not inherit permissions from it. It is up to the user to make an informed decision, case by case.

**Comment by bbaassssiiee** [comment]
For unarchive of tarfile a proper implementation of "mode: preserve" would be quite useful...

**Pull request merged**